### DISCUSSION TITLE
Strange behavior in deployed multi-pages app

### DISCUSSION THREAD

**Discussion by anquetos** [Avaiga/taipy#1378] [Q&A]
Hello,

We have deployed a multi-pages app on a Debian machine following this page : [https://docs.taipy.io/en/latest/manuals/run-deploy/deploy/linux/ubuntu/](https://docs.taipy.io/en/latest/manuals/run-deploy/deploy/linux/ubuntu/).

The app is online but we have strange behavior on pages where there are some controls (selectors).
To sum up (I could give details if needed), on one page we have a selector from which depends different values displayed in the rest of the page. This could be number, text, table or charts.

What we can see is the fact that sometimes values seem to change themselves or values are wrong (which is not the fact in local mode). So we think the state is shared between the users on the site at the same time which seems strange if we read [this](https://docs.taipy.io/en/latest/manuals/run-deploy/deploy/linux/ubuntu/).

The actual where it happens is called 'company' the pages is added to the Gui from the `main.py` file :
`gui_multi_pages = Gui(
    pages=pages,
    css_file="css/style.css"
)`

In the `company.py` file we initialize a few variables  where `selected_company` is the variable linked to the selector.
```
# Initialize selected company
selected_company = "SHELL"

# Create a filtered DataFrame with selected company
df_selected_company = data[data["mnc"] == selected_company]

# List companies to populate selector
colname_company = "mnc"
selector_company = list(np.sort(data[colname_company].astype(str).unique()))

# List years to populate selector and initialise selected year
selector_year = data.loc[data["mnc"] == selected_company, "year"].unique().astype(str).tolist()
selected_year = max(data.loc[data["mnc"] == selected_company, "year"].unique().tolist())

# Calculate number of reports for all companies
df_count_company = algo.number_of_tracked_reports_over_time_company(df_selected_company)

# Calculate sector, upe_code for selected company
company_sector = list(df_selected_company["sector"].unique())[0]
company_upe_name = df_selected_company["upe_name"].unique()[0]
```

In `company.md` : 
`<|{selected_company}|selector|lov={selector_company}|on_change=on_change_company|dropdown|class_name=fullwidth mb2|width=65%|>`

Then we use the `on_change_company` function to modify values : 

```
def on_change_company(state):
    print("Chosen company: ", state.selected_company)
    state.selected_company = state.selected_company

    # Update selected year and available years
    state.selector_year = data.loc[data["mnc"] == state.selected_company, "year"].unique().astype(str).tolist()
    state.selected_year = max(data.loc[data["mnc"] == state.selected_company, "year"].unique().tolist())
    print("Available years :", state.selected_year)

    state.df_selected_company = data[data["mnc"] == state.selected_company]
    state.df_count_company = algo.number_of_tracked_reports_over_time_company(state.df_selected_company)

    # Update content of page
    state.company_sector = list(state.df_selected_company["sector"].unique())[0]
    state.company_upe_name = state.df_selected_company["upe_name"].unique()[0]
```

So what did we do wrong ? Should we initialize the state of variables instead of initializing them directly ? Should initialze the state on the main ? We are a bit lost for the moment...

Thank you for your help.

**Comment by AlexandreSajus** [comment]
Hey, Alex from Taipy here!

This is a complex support request. It would be better to meet with us to debug this. Feel free to contact me at `alexandre.sajus@taipy.io` to schedule a call.

Something has to be changing the data instead of the `state.data` somewhere, which is changing the default value.

A good way to fix this would be to use `on_init(state)`, a Taipy callback that gets called when a new user connects to the app.

Try to add:

```python
def on_init(state):
  state.company_name = "SHELL"
  state.data = regenerate the data for SHELL
  state.viz = ...
  ...
```

Also, you can debug multi-user issues without deploying the app online: if you connect to your app in an incognito tab in your browser, it will simulate connecting with a new user

Hope this helps!

**Reply by anquetos** [reply]
Hello and thank you for your answer and proposition.

For the moment we managed to handle the situation with `on_init(state)` + an initialization of bindings variables used in the md files.

But I think the documentation is not precise enough about best practices to use `state` correctly. By reading the documentation we have been several people to think that there was nothing special to do about the state.

**Reply by FlorianJacta** [reply]
You are right on this issue. This is not really clear. This behavior does not occur when using basic Python types; it is seen when using more complex objects. Do you want to create [an issue](https://github.com/Avaiga/taipy-doc/issues/new?assignees=&labels=%F0%9F%93%88+Improvement&projects=&template=feature-improvement.md&title=) on our documentation to explain what would be the ideal solution? We can create one for you if you'd like

**Reply by anquetos** [reply]
Please can you create it for me ? Thank you.

**Reply by FlorianJacta** [reply]
https://github.com/Avaiga/taipy-doc/issues/989
